File: lib/core/utils/currency_formatter.dart
import 'package:intl/intl.dart';

/// Currency formatting — UI never shows raw doubles. BL-005.
class CurrencyFormatter {
  static final _standard = NumberFormat.currency(
    locale: 'tr_TR',
    symbol: '₺',
    decimalDigits: 2,
  );

  static final _compact = NumberFormat.compactCurrency(
    locale: 'tr_TR',
    symbol: '₺',
    decimalDigits: 1,
  );

  static final _noDecimal = NumberFormat.currency(
    locale: 'tr_TR',
    symbol: '₺',
    decimalDigits: 0,
  );

  /// Standard: ₺1.250,00
  static String format(double amount) => _standard.format(amount);

  /// No decimal: ₺1.250
  static String formatNoDecimal(double amount) => _noDecimal.format(amount);

  /// Compact: ₺1,2M for 1M+, ₺102K for 10K+, ₺1.250 for smaller
  static String compact(double amount) {
    final abs = amount.abs();
    if (abs >= 1000000) return _compact.format(amount);
    if (abs >= 10000) {
      final k = (amount / 1000).round();
      return '₺${k}K';
    }
    return formatNoDecimal(amount);
  }

  /// Signed: +₺1.250,00 or -₺500,00
  static String withSign(double amount) {
    final abs = format(amount.abs());
    return amount >= 0 ? '+$abs' : '-$abs';
  }

  /// Change percent: +%5,2 or -%3,1
  static String changePercent(double ratio) {
    final pct = (ratio * 100).abs();
    final sign = ratio >= 0 ? '+' : '-';
    return '$sign%${pct.toStringAsFixed(1)}';
  }

  /// Percent: %38,5
  static String percent(double ratio) =>
      '%${(ratio * 100).toStringAsFixed(1)}';

  /// Parse Turkish currency input to double
  static double? parse(String input) {
    final cleaned = input
        .replaceAll('₺', '')
        .replaceAll(' ', '')
        .replaceAll('.', '') // TR thousands separator
        .replaceAll(',', '.'); // TR decimal → dot
    return double.tryParse(cleaned);
  }
}

File: SavvyApp/Packages/Core/SavvyFoundation/Sources/SavvyFoundation/Formatters/CurrencyFormatter.swift
import Foundation

public enum CurrencyFormatter {
    private static let trLocale = Locale(identifier: "tr_TR")

    /// "₺1.250,00"
    public static func format(_ amount: Decimal) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "TRY"
        formatter.locale = trLocale
        return formatter.string(from: amount as NSDecimalNumber) ?? "₺0"
    }

    /// "₺1.250"
    public static func formatNoDecimal(_ amount: Decimal) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "TRY"
        formatter.locale = trLocale
        formatter.maximumFractionDigits = 0
        return formatter.string(from: amount as NSDecimalNumber) ?? "₺0"
    }

    /// "₺1,2M" / "₺102K" / "₺1.250"
    public static func compact(_ amount: Decimal) -> String {
        let d = NSDecimalNumber(decimal: amount).doubleValue
        switch abs(d) {
        case 1_000_000...:
            return String(format: "₺%.1fM", d / 1_000_000)
                .replacingOccurrences(of: ".", with: ",")
        case 10_000...:
            return "₺\(Int(d / 1_000))K"
        default:
            return formatNoDecimal(amount)
        }
    }

    /// "+₺1.250,00" / "-₺500,00"
    public static func withSign(_ amount: Decimal) -> String {
        let prefix = amount >= 0 ? "+" : ""
        return prefix + format(amount)
    }

    /// "+%5,2" / "-%3,1"
    public static func changePercent(_ ratio: Double) -> String {
        let sign = ratio >= 0 ? "+" : "-"
        let pct = abs(ratio) * 100
        return "\(sign)%\(String(format: "%.1f", pct).replacingOccurrences(of: ".", with: ","))"
    }

    /// "%38,5"
    public static func percent(_ ratio: Double) -> String {
        "%\(String(format: "%.1f", ratio * 100).replacingOccurrences(of: ".", with: ","))"
    }

    /// Reverse Turkish format → Decimal
    public static func parse(_ input: String) -> Decimal? {
        let cleaned = input
            .replacingOccurrences(of: "₺", with: "")
            .replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: ".", with: "")
            .replacingOccurrences(of: ",", with: ".")
            .trimmingCharacters(in: .whitespaces)
        return Decimal(string: cleaned)
    }
}
